File: SDE_sheet/Arrays/02_pascal_triangle/solution.py
class Solution:
    def generate(self, numRows: int) -> List[List[int]]:
        nums = []

        for i in range(numRows):
            temp = []

            for j in range(i + 1):
                if j == 0 or j == i:
                    temp.append(1)
                else:
                    prevRow = nums[i - 1]
                    left = prevRow[j - 1] if j - 1 >= 0 else 0
                    right = prevRow[j] if j < len(prevRow) else 0
                    temp.append(left + right)

            nums.append(temp)

        return nums

File: SDE_sheet/Arrays/02_pascal_triangle/solution.java
class Solution {
    public List<List<Integer>> generate(int numRows) {
        List<List<Integer>> nums = new ArrayList<>();

        for (int i = 0; i < numRows; i++) {
            List<Integer> temp = new ArrayList<>();

            for (int j = 0; j <= i; j++) {
                if (j == 0 || j == i) {
                    temp.add(1);
                } else {
                    List<Integer> prevRow = nums.get(i - 1);
                    int left = prevRow.get(j - 1);
                    int right = prevRow.get(j);
                    temp.add(left + right);
                }
            }

            nums.add(temp);
        }

        return nums;
    }
}
